File: january_2023/CircularDeque.java
class MyCircularDeque {
    int size;
    int maxsize;
    List<Integer> queue;

    public MyCircularDeque(int k) {
        this.maxsize = k;
        this.size = 0;
        this.queue = new ArrayList<>();
    }
    
    public boolean insertFront(int value) {
        if (! this.isFull()){
            this.queue.add(0, value);
            this.size += 1;
            return true;
        }
        return false;
    }
    
    public boolean insertLast(int value) {
        if (! this.isFull()){
            if (! this.isEmpty()){
                this.queue.add(this.size, value);
                this.size += 1;
                return true;
            }
            else {
                return this.insertFront(value);
            }
        }    
        return false;
    }
    
    public boolean deleteFront() {
        if (! this.isEmpty()){
            this.queue.remove(0);
            this.size -= 1;
            return true;
        }
        return false;
    }
    
    public boolean deleteLast() {
        if (! this.isEmpty()){
            this.queue.remove(this.size - 1);
            this.size -= 1;
            return true;
        }
        return false;
    }
    
    public int getFront() {
        if (! this.isEmpty()){
            return this.queue.get(0);
        }
        return -1;
    }
    
    public int getRear() {
        if (! this.isEmpty()){
            return this.queue.get(this.size - 1);
        }
        return -1;
    }
    
    public boolean isEmpty() {
        return this.size == 0;
    }
    
    public boolean isFull() {
        return this.size == this.maxsize;
    }
}

File: january_2023/CircularDeque.py
class MyCircularDeque:

    def __init__(self, k: int):
        self.queue = []
        self.size = 0
        self.maxSize = k

    def insertFront(self, value: int) -> bool:
        if not self.isFull():
            self.queue.insert(0, value)
            self.size += 1
            return True
        return False

    def insertLast(self, value: int) -> bool:
        if not self.isFull():
            self.queue.append(value)
            self.size += 1
            return True
        return False

    def deleteFront(self) -> bool:
        if not self.isEmpty():
            self.queue = self.queue[1:]
            self.size -= 1
            return True
        return False

    def deleteLast(self) -> bool:
        if not self.isEmpty():
            self.queue = self.queue[:self.size - 1]
            self.size -= 1
            return True
        return False

    def getFront(self) -> int:
        if not self.isEmpty():
            return self.queue[0]
        return -1

    def getRear(self) -> int:
        if not self.isEmpty():
            return self.queue[self.size - 1]
        return -1

    def isEmpty(self) -> bool:
        return self.size == 0
        
    def isFull(self) -> bool:
        return self.size == self.maxSize
